### PR TITLE
Feat/etcd leader

### PR DESCRIPTION
**Pull request by nioshield**
Closes: #91

### CODE REVIEW

**Comment by shafreeck** [comment]
Nice improvement! I love the idea to use Etcd to do the election.

However, I have some different ideas about the design of the election flow.

## Use the synchronous way to run an election

What you did in this PR is in an asynchronous way, which is:
1. Register to campaign the leader
2. Set the leader flag as true
3. Check the flag in a loop and call `runExpire` in the loop.

I prefer to use the synchronous way in fact, which is:
1. Campaign the leader and block until it is elected
2. Start the expiration goroutine, and keeps watching on the session's leadership
3. Exit the goroutine and return to run Campaign if the session's leadership is lost.

Describe in the pseudo-code

```
for{
 sess = NewSession
 elec = NewElection(sess)
 elec.Campaign() // block until being elected
 StartExpire(sess)
}

func StartExpire(sess){
  for {
    select {
      case sess.Done():
        return
    }
  ...
 }
}
```

## Avoid using a name like xxxutil

These general words like `util`, `common` should be avoided as can as possible. After refactoring like above, `etcdutil` is probably not used anymore. If it is needed, use the name to describe the feature like `election`, if you really need to wrap etcd as a thin layer, use the name `etcd` directly.

**Comment by arthurkiller** [comment]
agreed

File: db/etcdutil/elect.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+		key: key,
+		val: val,
+		ttl: ttl,
+	}
+	go e.Campaign()
+	return e
+}
+
+func (e *Elect) Campaign() {
+	for {
```
Comment (arthurkiller) [inline db/etcdutil/elect.go:35]: 这个 for 做的事情有点多， 感觉全哥说的那个方式挺好的

Comment (arthurkiller) [reply db/etcdutil/elect.go:35]: 即：
1. for{
campaign()
}
2. run expire

**Review by arthurkiller — APPROVED** [review]
LGTM

**Pull request merged**